tuna/machine.py

```diff
@@ -83,63 +92,73 @@ def __init__(self, **kwargs):
     self.__dict__.update(
         (key, value) for key, value in kwargs.items() if key in allowed_keys)
 
-    self.cnx_list = {}
-    self.log_list = {}
-    self.num_cpus = 0
+    self.cnx_list: dict = {}
+    self.log_list: dict = {}
+    self.num_cpus: int = 0
+    self.avail_gpus: List[Dict[int, str]]
+    self.sclk: int
+    self.mclk: int
+    self.gpus: List[Dict[Any, Any]]
+    self.cpus: List[dict]
+    self.logger: logging.Logger
+    self.mmi: MachineManagementInterface
+    self.cnx: Connection
+    self.ssh: paramiko.SSHClient = SSHClient()
+
     if self.local_machine:  # pylint: disable=no-member ; false alarm
       self.logger = setup_logger(f'Machine_{self.hostname}')
       self.connect()
       self.mmi = None
-      self.id = 0  # pylint: disable=invalid-name
+
+      self.id: int = 0  # pylint: disable=invalid-name
       self.get_properties()
 
       if self.gpus:
         self.arch = self.gpus[0]['arch']
         self.num_cu = self.gpus[0]['num_cu']
-        self.avail_gpus = list(range(len(self.gpus)))
+        self.avail_gpus = list(range(len(self.gpus)))  #type: ignore
     else:
-      cmd = 'hostname'
+      cmd: str = 'hostname'
       with Popen(cmd, stdout=PIPE, shell=True, universal_newlines=True) as subp:
-        hostname = subp.stdout.readline().strip()
+        if subp.stdout is not None:
+          hostname: str = subp.stdout.readline().strip()
       if self.local_ip is not None and check_qts(hostname):
         self.hostname = self.local_ip
-        self.port = self.local_port
+        self.port: int = self.local_port
       self.logger = setup_logger('Machine_{self.id}')
       self.mmi = MachineManagementInterface(self.ipmi_ip, self.ipmi_port,
                                             self.ipmi_user, self.ipmi_password)
 
       if not self.avail_gpus is None:
-        self.avail_gpus = [int(val) for val in self.avail_gpus.split(',')]
+        self.avail_gpus = [
+            int(val) for val in self.avail_gpus.split(',')  #type: ignore
+        ]  #type: ignore
         self.num_gpus = len(self.avail_gpus)
```

Comment:
avail_gpus could come in through kwargs as a str. So it's ok to ignore that type complication here. This input parsing should probably be moved outside of this constructor in a separate PR.

Reply:
Ok. separate PR is ideal.